fix(startup): fix start standalone outside vscode

binhtran432k opened the pull request:
<!---
Thanks for helping to make Cucumber better! 💖

You can feel free to open a "Draft" status pull request if you're not ready for feedback yet.

Don't worry about getting everything perfect! We're here to help and will coach you through
to getting your pull request ready to merge.

The prompts below are for guidance to help you describe your change in a way that is most 
likely to make sense to other people when they are reviewing it. Still, it's just a guide, 
so feel free to delete anything that doesn't feel appropriate, and add anything additional 
that seems like it would provide useful context. 👏🏻
-->

### 🤔 What's changed?

<!-- Describe your changes in detail -->

### ⚡️ What's your motivation? 

<!-- 
What motivated you to propose this change? Does it fix a bug? Add a new feature?
If it fixes an open issue, you can link to the issue here, e.g. "Fixes #99"
-->
Fixes #72, #79, #90, #96

### 🏷️ What kind of change is this?

<!--- Delete any options that are not relevant -->

- :bug: Bug fix (non-breaking change which fixes a defect)

### ♻️ Anything particular you want feedback on?

<!-- 
Is there anything in this change you're unsure about, or would 
particularly like reviewers to give you feedback on?
-->

First, The `url.pathToFileURL` was introduced by #68 to fix #66 but it did not fix the problem and #66 was closed by using another product.

After that, the #73 change how `language-server` worked. It refactor `src/wasm/startWasmServer.ts` into `src/wasm/startEmbeddedServer.ts` and `src/wasm/startStandaloneServer.ts`. The `startEmbeddedServer` is used for VS Code with browser support and the `startStandaloneServer` is used as fallback for user outside of VS Code but it was not tested properly.

As a result, the language server is rendered unusable outside of VS Code, and this PR aims to rectify this issue.

### 📋 Checklist:

<!--- 
This is to help you remember all the little things we often forget to do!

Feel free to delete any tasks that are not relevant, or add new ones.
-->

- [x] I agree to respect and uphold the [Cucumber Community Code of Conduct](https://cucumber.io/conduct/)
- [x] I've changed the behaviour of the code
  - [x] I have added/updated tests to cover my changes.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.
- [x] Users should know about my change
  - [x] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../CHANGELOG.md), linking to this pull request.

----

*This text was originally generated from a [template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates), then edited by hand. [You can modify the template here.](https://github.com/cucumber/.github/edit/main/.github/PULL_REQUEST_TEMPLATE.md)*

pprotas commented:
Can you provide a guide on how to install this branch locally for my neovim config? Seems like it takes too long for the maintainer to merge it

noah-friedman commented:
This also fixes #79

noah-friedman commented:
> Can you provide a guide on how to install this branch locally for my neovim config? Seems like it takes too long for the maintainer to merge it

@pprotas `npm i -g @binhtran432k/cucumber-language-server@1.4.0`

lkoning commented:
hi all, thank you for your work! I tried to use your lsp-version in my Lazyvim by your instructions:

1. Install the lsp locally - :white_check_mark: 
2. Config Lazyvim:
```lua
return {
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        cucumber_language_server = {
          -- currently using a forked ls. 
          mason = false,
          root_dir = require("lspconfig.util").root_pattern("pom.xml"),
          settings = {
            cucumber = {
              features = { "**/*.feature" },
              glue = { "**/*Steps.java" },
            },
          },
        },
      },
    },
  },
}
```
Now, after loading my Java repo, or a JS repo...LSP seems to be active but formatting and finding stepDefs does not work. Haven't tried Vscode. 

LspLog:
```log
[START][2023-11-14 10:18:12] LSP logging initiated
[WARN][2023-11-14 10:18:12] ...lsp/handlers.lua:137	"The language server cucumber_language_server triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
[ERROR][2023-11-14 10:18:12] ...lsp/handlers.lua:535	"Failed to reindex: Cannot read properties of undefined (reading 'reduce')"
[ERROR][2023-11-14 10:18:12] ...lsp/handlers.lua:535	"Failed to reindex: Cannot read properties of undefined (reading 'reduce')"
``` 

I don't really know what this means. At least it seems we are a few steps further.

shlomisaad commented:
> hi all, thank you for your work! I tried to use your lsp-version in my Lazyvim by your instructions:
> 
> 1. Install the lsp locally - ✅
> 2. Config Lazyvim:
> 
> ```lua
> return {
>   {
>     "neovim/nvim-lspconfig",
>     opts = {
>       servers = {
>         cucumber_language_server = {
>           -- currently using a forked ls. 
>           mason = false,
>           root_dir = require("lspconfig.util").root_pattern("pom.xml"),
>           settings = {
>             cucumber = {
>               features = { "**/*.feature" },
>               glue = { "**/*Steps.java" },
>             },
>           },
>         },
>       },
>     },
>   },
> }
> ```
> 
> Now, after loading my Java repo, or a JS repo...LSP seems to be active but formatting and finding stepDefs does not work. Haven't tried Vscode.
> 
> LspLog:
> 
> ```
> [START][2023-11-14 10:18:12] LSP logging initiated
> [WARN][2023-11-14 10:18:12] ...lsp/handlers.lua:137	"The language server cucumber_language_server triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
> [ERROR][2023-11-14 10:18:12] ...lsp/handlers.lua:535	"Failed to reindex: Cannot read properties of undefined (reading 'reduce')"
> [ERROR][2023-11-14 10:18:12] ...lsp/handlers.lua:535	"Failed to reindex: Cannot read properties of undefined (reading 'reduce')"
> ```
> 
> I don't really know what this means. At least it seems we are a few steps further.

I have the same problem myself, I also get Undefined step everywhere

M4urici0GM commented:
Using the branch `npm i -g @binhtran432k/cucumber-language-server@1.4.0` i was able to configure my lsp with config:
```lua
{
    cmd = { "~/.nvm/versions/node/v16.20.2/bin/cucumber-language-server", "--stdio" },
    settings = {
        cucumber = {
            cucumber = {
                features = { "**/*.feature" },
                glue = { "**/*Step*.java" },
            }
        }
    },
    on_attach = function(client, bufnr)
        print("attached to cucumber file")
        vim.keymap.set('n', "gd", vim.lsp.buf.definition, { buffer = 0 })
        vim.keymap.set('n', "gn", vim.diagnostic.goto_next, { buffer = 0 })
        vim.keymap.set('n', "gb", vim.diagnostic.goto_prev, { buffer = 0 })
    end
}

kieran-ryan commented on bin/cucumber-language-server.cjs:
```diff
@@ -4,12 +4,15 @@ require('source-map-support').install()
   `${__dirname}/../node_modules/@cucumber/language-service/dist`
 )
-const { connection } = startStandaloneServer(wasmBaseUrl.href, (rootUri) => new NodeFiles(rootUri))
+const { connection } = startStandaloneServer(
+  url.fileURLToPath(wasmBaseUrl),
+  (rootUri) => new NodeFiles(rootUri)
+)
```
Can you provide more information on rationale to change from a `file:///` URL to a file path? I'm also wondering would there be impact if we wanted to run the language server in a browser-based context in future.

- Perhaps, is there any error output you can provide that indicates the reason for this change?

If we are no longer using the URL, and simply want to use a file path; then I would consider, can we use `path` instead? An improvement in any case would be to avoid calling functions inside the arguments of another function where possible, which can make it easier to debug based on line number of failures e.g. assign`url.fileURLToPath(wasmBaseUrl)` to a variable, and then reference it from function call 👍 Below is just a quick reference

```javascript
const path = require('path');
const distPath = path.resolve(`${__dirname}/../node_modules/@cucumber/language-service/dist`)
const { connection } = startStandaloneServer(distPath, (rootUri) => new NodeFiles(rootUri))
```

Has this been tested on Windows? Has it been tested through VSCode - see [contributing guidelines](https://github.com/cucumber/vscode/blob/main/CONTRIBUTING.md)? Would like to ensure works with Windows paths. Are there any unit tests that could be added for this module that could provide us with more confidence in changes to this file now and going forward?

binhtran432k replied:
Hi @kieran-ryan,

Sorry about the delay in my response. I don't use Windows regularly, so it took me some time to set up an environment for testing. I've now tested the PR on Windows, but only with Node.js v16.x.x. This is due to compatibility issues with the latest release of the `language-client` package on newer Node.js versions.

I've also updated the PR description to provide more context on the reasons behind these changes. Please refer to the updated description for a clearer understanding of their value.

Thank you for your patience and understanding.

binhtran432k replied:
Hi @kieran-ryan,

I'd suggest omitting setup documentation for Neovim users within this PR, as it's already comprehensively covered in the Neovim community documentation at [https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#cucumber_language_server](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#cucumber_language_server).

Regarding unit testing, effectively testing this functionality requires a real LSP client. I'll address this in a separate PR, if deemed necessary, to ensure thorough testing.

kieran-ryan reviewed (commented):
Firstly, thanks @binhtran432k for making an excellent contribution to the project - and welcome to our [Slack community](https://cucumber.io/community#slack) also; secondly thanks all for your patience on this - a review has been pending for a considerable time.

I'd like to support to get this reviewed and merged. However, I'm unfamiliar with Neovim - so you might have to bear with me while I get to grips with spinning up the Language Server with it, hopefully end of next week.

To close this out, some items we may need to achieve are as follows:

- [x] Tested in VSCode - see [contributing guidelines](https://github.com/cucumber/vscode/blob/main/CONTRIBUTING.md)
- [ ] Tested on Windows - particularly the file path change
- [ ] Consider whether unit tests can be written covering the file path change - ideally tests that fail with the previous state and pass with the changed state
- [ ] Ensure changing from `file:///` to a path would not have any impact if we were to enable the language server to work in browser-based contexts in future
- [ ] Consider whether we should include a lightweight `Usage` or `Installation` section in the `README` with a subsection for using Neovim - explaining how to complete installation - perhaps this is an area you could assist @harleyreevesmartin - as a newer user looking to contribute and as I observe you are a Neovim user

binhtran432k commented on bin/cucumber-language-server.cjs:
```diff
@@ -4,12 +4,15 @@ require('source-map-support').install()
 const { NodeFiles } = require('../dist/cjs/src/node/NodeFiles')
 const url = require('url')
-const { version } = require('../src/version')
+const { version } = require('../dist/cjs/src/version')
```
> Can you advise on whether this change is required? Or was just made during local development?

I just use the same strategy as `NodeFiles` on line 5. You can run `node ./bin/cucumber-language-server.cjs --stdio` on the main branch and this branch to see the effect.

binhtran432k commented on src/wasm/startStandaloneServer.ts:
```diff
@@ -12,4 +12,8 @@ export function startStandaloneServer(wasmBaseUrl: string, makeFiles: (rootUri:
   new CucumberLanguageServer(connection, documents, adapter, makeFiles, () => undefined)
   connection.listen()
+
+  return {
+    connection,
+  }
```
> Appears we were assigning the output of `startStandaloneServer` to `connection`, but were simply not returning the connection itself to it. Can you confirm that is accurate?
> 
> ```js
> const { connection } = startStandaloneServer(wasmBaseUrl.href, (rootUri) => new NodeFiles(rootUri))
> ```
Yes, I've confirmed that the connection itself is no longer being returned directly from startStandaloneServer. I reviewed the previous version of the file and verified that it did return the connection previously. However, this behavior changed after the language server was split into embedded and standalone modes.

kumpan-david commented:
I managed to get it working with a little hacking using `nvim v0.9.1` on macOS Sonoma 14.2.1.
Although I only tested with `@cucumber/language-server@1.2.0` since that's the version the `Mason` Registry provides.

Since `tree-sitter` [currently crashes on node > 19](https://github.com/tree-sitter/tree-sitter/issues/2338) I reinstalled the dependencies using node v16.17.0 and patched the bin-script to use the node-server instead of the wasm-server.

So if anyone using `nvim`, `lspconfig` and `Mason` comes, here this is a temporary workaround.

```js
//cucumber-language-server/node_modules/.bin/cucumber-language-server
#!/change/this/path/to/v16.17.0/bin/node
/* eslint-disable @typescript-eslint/no-var-requires */
require('source-map-support').install()
const url = require('url')
const { startNodeServer, NodeFiles } = require('@cucumber/language-server/node')
const wasmBaseUrl = url.pathToFileURL(
  `file:///${__dirname}/../node_modules/@cucumber/language-service/dist`
)

startNodeServer(wasmBaseUrl.href, (rootUri) => new NodeFiles(rootUri))

```

Another detail I had to add was to double up the `lspconfig` settings and pass empty parameterTypes to prevent some errors from `@cucumber/language-service`
```lua
require("lspconfig").cucumber_language_server.setup({
  settings = {
    cucumber = {
      features = { "**/*.feature" },
      glue = { "**/*.steps.ts" },
      parameterTypes = {},
    },
    features = { "**/*.feature" },
    glue = { "**/*.steps.ts" },
    parameterTypes = {},
  },
})


```

binhtran432k commented:
> I managed to get it working with a little hacking using `nvim v0.9.1` on macOS Sonoma 14.2.1. Although I only tested with `@cucumber/language-server@1.2.0` since that's the version the `Mason` Registry provides.
> 
> Since `tree-sitter` [currently crashes on node > 19](https://github.com/tree-sitter/tree-sitter/issues/2338) I reinstalled the dependencies using node v16.17.0 and patched the bin-script to use the node-server instead of the wasm-server.
> 
> So if anyone using `nvim`, `lspconfig` and `Mason` comes, here this is a temporary workaround.
> 
> ```js
> //cucumber-language-server/node_modules/.bin/cucumber-language-server
> #!/change/this/path/to/v16.17.0/bin/node
> /* eslint-disable @typescript-eslint/no-var-requires */
> require('source-map-support').install()
> const url = require('url')
> const { startNodeServer, NodeFiles } = require('@cucumber/language-server/node')
> const wasmBaseUrl = url.pathToFileURL(
>   `file:///${__dirname}/../node_modules/@cucumber/language-service/dist`
> )
> 
> startNodeServer(wasmBaseUrl.href, (rootUri) => new NodeFiles(rootUri))
> ```
> 
> Another detail I had to add was to double up the `lspconfig` settings and pass empty parameterTypes to prevent some errors from `@cucumber/language-service`
> 
> ```lua
> require("lspconfig").cucumber_language_server.setup({
>   settings = {
>     cucumber = {
>       features = { "**/*.feature" },
>       glue = { "**/*.steps.ts" },
>       parameterTypes = {},
>     },
>     features = { "**/*.feature" },
>     glue = { "**/*.steps.ts" },
>     parameterTypes = {},
>   },
> })
> ```

Hi @kumpan-david, thank you again for your help. Unfortunately, a significant refactor implemented after v1.2.0 makes reverting to that version impractical. However, I'm glad to report that this fix works with the latest version. I appreciate your understanding.

binhtran432k commented:
Hi @kieran-ryan,

I've pushed new commits that accomplish the following:

**Testing:**

- Completed testing in VSCode.
- Conducted testing on Windows.

**Unit Testing:**

- Added a unit test that starts the actual language server and verifies the absence of errors.
- Noted a minor error (out of scope for this fix) with a TODO in the test file.

**External VSCode Usage:**

- Added a note for users working outside of VSCode.

Please review the changes and let me know if you have any questions or feedback.

FuzzyAzurik commented:
@binhtran432k from a fellow neovim user thank you so much!
I was able to get a working cucumber-language-server by installing your branch!
I am using node 18.19.0 and neovim 0.9.5

This issue has been bugging me for months!

helgardferreira commented:
Any chance this can get merged anytime soon?

kieran-ryan commented on test/standalone.test.ts:
```diff
@@ -0,0 +1,144 @@
+  })
+})
+
+class NullLogger implements Logger {
+  error(): void {
+    // no-op
+  }
+
+  warn(): void {
+    // no-op
+  }
+
+  info(): void {
+    // no-op
+  }
+
+  log(): void {
+    // no-op
+  }
+}
```
As this code is shared with `CucumberLanguageServer.test.ts`, would it be possible to extract the function into a separate module and import to both?

https://github.com/cucumber/language-server/blob/c2fbe5f66112d2e25df252094ecd8182c7bacf7b/test/CucumberLanguageServer.test.ts#L210-L226

kieran-ryan reviewed (changes requested):
Excellent work @binhtran432k - appreciate the changes. Just two small comments to clarify the README documentation and to reuse some test code, and then we can merge and go for release. Well done! 👏

binhtran432k commented:
@kieran-ryan, Thank you for your review. I have just completed the requested changes.

kieran-ryan commented:
[v1.5.0](https://www.npmjs.com/package/@cucumber/language-server/v/1.5.0) is now available 🚀
